triedb/pathdb/difflayer.go

```diff
@@ -133,7 +221,32 @@ func (dl *diffLayer) node(owner common.Hash, path []byte, hash common.Hash, dept
 // Node implements the layer interface, retrieving the trie node blob with the
 // provided node information. No error will be returned if the node is not found.
 func (dl *diffLayer) Node(owner common.Hash, path []byte, hash common.Hash) ([]byte, error) {
-	return dl.node(owner, path, hash, 0)
+	if n := dl.cache.Get(hash); n != nil {
+		// The query from the hash map is fastpath,
+		// avoiding recursive query of 128 difflayers.
+		diffHashCacheHitMeter.Mark(1)
+		diffHashCacheReadMeter.Mark(int64(len(n.Blob)))
+		return n.Blob, nil
+	}
+	diffHashCacheMissMeter.Mark(1)
+
+	parent := dl.parent
+	for {
+		if disk, ok := parent.(*diskLayer); ok {
+			blob, err := disk.NodeByLogger(owner, path, hash, log.Debug)
+			if err != nil {
+				// This is a bad case with a very low probability. The same trienode exists
+				// in different difflayers and can be cleared from the map in advance. In
+				// this case, the 128-layer difflayer is queried again.
+				diffHashCacheSlowPathMeter.Mark(1)
```

Comment:
I think we should add ref for every node in cache to avoid this bad case.

Reply:
Updated.